src/middleware/authUser.js

```diff
@@ -0,0 +1,36 @@
+const boom = require('boom')
+
+const roles = require('../constants/roles.js')
+const { auth } = require('../constants/errors.json')
+
+const User = require('../models/User')
+
+// function to check role of user matches minRole allowed on the route
+const verifyRole = ({ minRole }) => user => {
+  const userScope = user.scope
+
+  if (minRole === roles.BASIC) return
+  if (minRole === roles.ADMIN && userScope !== roles.BASIC) return
+  if (minRole === roles.SUPER && userScope === roles.SUPER) return
+
+  return Promise.reject(boom.unauthorized(auth.UNAUTHORIZED))
+}
+
+// Check user from the token exists, and add their id to the req.user for potential use in permissioning
+const addUserIdToSession = req => {
+  return User.find({ username: req.user.username }).then(users => {
+    if (users.length === 0) {
+      return Promise.reject(boom.unauthorized(auth.UNAUTHORIZED))
+    }
+
+    req.user.id = users[0].id
+    return req.user
+  })
+}
+
+module.exports = opts => (req, res, next) => {
+  addUserIdToSession(req)
```

Comment:
@m4v15 

  1. We can add the username to the jwt.
  2. We need the id to check the permission, ie `PUT events/idofeoinsevent`, that need the id.
  3. We want the same user/session/auth object attached to the request, in both the oauth and the cookie/jwt case, its not too important how we get there

This is the flow I see

Request -> auth -> permission -> controller
auth: verifies the identity of the user, ie checks cookie / auth header and attaches user object to request (id, role, username ..)
permission: optional, if present asserts on the user object. ie: must be logged in; or, must have role SUPER; or must have same id as owner of requested resource

Does that make sense / bring clarity?